src/client.rs

```diff
@@ -141,6 +142,72 @@ impl<IO> AsyncWrite for TlsStream<IO>
 where
     IO: AsyncRead + AsyncWrite + Unpin,
 {
+    fn poll_write_vectored(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &[IoSlice<'_>],
+    ) -> Poll<Result<usize, Error>> {
+        let this = self.get_mut();
+        let mut stream =
+            Stream::new(&mut this.io, &mut this.session).set_eof(!this.state.readable());
+
+        #[allow(clippy::match_single_binding)]
+        match this.state {
+            #[cfg(feature = "early-data")]
+            TlsState::EarlyData(ref mut pos, ref mut data) => {
+                use std::io::Write;
+
+                // write early data
+                if let Some(mut early_data) = stream.session.early_data() {
+                    let mut written = 0;
+
+                    for buf in bufs {
+                        let len = early_data.write(buf)?;
+                        data.extend_from_slice(&buf[..len]);
+
+                        // If we didn't write the entire buffer, return early.
+                        if len != buf.len() {
+                            return Poll::Ready(Ok(written));
+                        }
```

Comment:
I am not sure this is entirely correct. We will be called again because we returned `Poll::Ready`, ending up in an endless loop because `len` will probably be `0` and we thus won't ever get out of this.

Perhaps we shouldn't bother with looping over `bufs` and just only write the first non-empty buffer? As long as we return `Poll::Ready`, any caller _should_ call us again until we return `Poll::Pending`.

Reply:
If early_data write returns 0, it means the early_data buffer is full and we should break out of loop and try handshake.

I agree that it is safe to just write to first non-empty buf and have no preference for this.